### PR TITLE
Add/user tracking option api

### PR DESCRIPTION
**Pull request by enejb**
#### Changes proposed in this Pull Request:

* Add an interface class to update if a user should be tracked. 
* Sync the data to .com so that we can stop tracking the users there as well when dey navigate around the site.

#### Testing instructions:
* Do the tests pass?

### CODE REVIEW

File: _inc/lib/class.jetpack-user-event-tracking.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	}
+
+	static function enable( $user_id ) {
+		// user opted in
+		self::set( $user_id, 1 );
+	}
+
+	static private function set( $user_id, $value ) {
+		self::$_cache[ $user_id ] = (bool) $value;
+		update_user_meta( $user_id, self::KEY, $value );
```
Comment (eliorivero) [inline _inc/lib/class.jetpack-user-event-tracking.php]: Please update this to return the `update_user_meta` and also return the `self::set` in `disable` and `enable`.

File: _inc/lib/class.jetpack-user-event-tracking.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	static function is_enabled( $user_id ) {
+		if ( isset( self::$_cache[ $user_id ] ) ) {
+			return self::$_cache[ $user_id ];
+		}
+		$user_tracking = get_user_meta( $user_id, self::KEY, true );
+		if ( ! is_numeric( $user_tracking ) ) {
+			$user_tracking = self::default_value();
+		}
+		self::$_cache[ $user_id ] = (bool) $user_tracking;
+		return (bool) $user_tracking();
```
Comment (eliorivero) [inline _inc/lib/class.jetpack-user-event-tracking.php]: Should be `$user_tracking`.

File: _inc/lib/class.jetpack-user-event-tracking.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
+<?php
+
+class Jetpack_User_Event_Tracking {
+
+	private static $_cache = array();
+	const KEY = 'jetpack_event_tracking';
```
Comment (eliorivero) [inline _inc/lib/class.jetpack-user-event-tracking.php]: This is not PHP 5.2 safe.

>or by using the const keyword outside a class definition as of PHP 5.3.0

http://php.net/language.constants.syntax

File: _inc/lib/class.jetpack-user-event-tracking.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	}
+
+	static function enable( $user_id ) {
+		// user opted in
+		return self::set( $user_id, 1 );
+	}
+
+	static private function set( $user_id, $value ) {
+		self::$_cache[ $user_id ] = (bool) $value;
+		return update_user_meta( $user_id, self::$KEY, $value );
```
Comment (eliorivero) [inline _inc/lib/class.jetpack-user-event-tracking.php]: Why are you casting this as boolean when you save it to the cache, but not when you update the user meta data?

Comment (enejb) [reply _inc/lib/class.jetpack-user-event-tracking.php]: Because I want to be able to distinguish between a set value (0,1) vs the default value (false)

File: _inc/lib/class.jetpack-user-event-tracking.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	}
+
+	static function default_value() {
+		/**
+		 * Return the default jetpack user event tracking opt out value.
+		 *
+		 * @since 6.0.0
+		 *
+		 * @param bool Default to false. (user tracking enabled)
```
Comment (eliorivero) [inline _inc/lib/class.jetpack-user-event-tracking.php]: Update this line to 
```
* @param bool $disable_tracking Whether to disable user tracking. Defaults to false, which allows user tracking.
```

File: tests/php.multisite.xml [hunk 6]
```diff
@@ -14,6 +14,7 @@
             <file>php/test_class.jetpack-client-server.php</file>
             <file>php/test_class.jetpack-xmlrpc-server.php</file>
             <file>php/test_class.jetpack-heartbeat.php</file>
+			<file>tests/php/test_class.jetpack-user-event-tracking.php</file>
```
Comment (eliorivero) [inline tests/php.multisite.xml]: Check indentation.

File: _inc/lib/class.jetpack-user-event-tracking.php (hunk 4 above)
Comment (eliorivero) [reply _inc/lib/class.jetpack-user-event-tracking.php]: I see, thanks!

**Comment by bperson** [comment]
👋 not sure how the sync to WP.com works: can it potentially be both ways? ( If someone sets it on WP.com via Calypso, have it be synced in Jetpack? )

For now I've been implementing the Calypso stuff using D9412-code as the backend.

File: _inc/lib/class.jetpack-user-event-tracking.php [hunk 7]
```diff
@@ -0,0 +1,58 @@
+<?php
+
+class Jetpack_User_Event_Tracking {
+
+	private static $_cache = array();
```
Comment (mattwiebe) [inline _inc/lib/class.jetpack-user-event-tracking.php]: Drive-by comment since I'm looking at this in concert with tracking on the Calypso side: what's the rationale behind having a local cache for this user meta? The user meta functions already leverage WP's built-in `wp_cache_*` functions, which also persist in memory for a session like this is doing.

Comment (enejb) [reply _inc/lib/class.jetpack-user-event-tracking.php]: The rational was that is would be faster. But I see your point and would be happy to remove it.

**Comment by enejb** [comment]
Hi @bperson 
We talked about this on our meet up when we were implementing this. The conclusion was that things get pretty complicated quickly and that each the user should be responsible for update the setting on each of the sites. 

>If someone sets it on WP.com via Calypso, have it be synced in Jetpack? 

Right now this wouldn't be the case that things get synced. We do however send this data to .com when the user sets it on the Jetpack side so that we can stop tracking the user's actions on a per site basis. 

For example currently we also do not sync the user's display name across the different Jetpack sites.
